### ISSUE TITLE
[Bug] MS Teams + MS Calendar Bad Implementation

### ISSUE THREAD

**Issue by ldpg-dev** [calcom/cal.com#14765]
### Issue Summary

Microsoft Teams + Microsoft Outlook Calendar Meetings are not being created as online meetings, and have unprofessional description.

### Steps to Reproduce

1. Install MS Teams app from the app marketplace
2. Setup MS Outlook as your main Calendar
3. Create a Booking and set MS Teams as the meeting software
4. Book a meeting
5. Atendees will receive 2 emails: 1 from Cal.com + 1 from MS Calendar

### Actual Results

- Attendees will get 2 emails. 1 from Cal.com and 1 from MS Calendar
- The one from MS Calendar looks like this:
<img width="919" alt="image" src="https://github.com/calcom/cal.com/assets/81204440/f07a8f6b-b52c-4241-a501-a4d9e8cb1eeb">

- When I go to MS Teams to check this is what I see:
<img width="644" alt="image" src="https://github.com/calcom/cal.com/assets/81204440/dd5d9f82-cc75-4ab6-b3be-878c0058c564">

### Expected Results

- This is what that email would normally look like when I create a meeting through Calendly or MS Teams directly:
<img width="557" alt="image" src="https://github.com/calcom/cal.com/assets/81204440/a3185eb4-a9f8-4089-9f0c-0bc4ea75c9a0">

-
_Notice how the location is Not the Meeting Link but an arbitrary URL, in this case Calendly's event, and the "Online Meeting" switch is turned on, letting Microsoft handle the link creation, instead of creating a link and then attaching it to the calendar event_
<img width="632" alt="image" src="https://github.com/calcom/cal.com/assets/81204440/03350cdf-01e6-42a1-b104-b7c0cdc3965f">

**Comment by milospuac** [comment]
@joeauyeung

**Comment by buzzard10** [comment]
Is this matter of permissions? Or bad implementation?

**Comment by ldpg-dev** [comment]
@buzzard10 This is a bad implementation of the MS Teams integration with Outlook/MS Calendar

**Comment by buzzard10** [comment]
@ldpg-dev, it was functioning for me a few weeks ago; I had a link to the meeting, but now it's not working. What has changed?

**Comment by ldpg-dev** [comment]
I honestly don't know, @milospuac and @joeauyeung have you been able to take a look at this? @ciaranha Not sure why it's not labeled as a bug anymore.

**Comment by krakowski** [comment]
The Microsoft Graph API doesn't support suppressing invitation mails when creating an [`event`](https://learn.microsoft.com/de-de/graph/api/resources/event) object.

  * https://techcommunity.microsoft.com/t5/microsoft-365-developer-platform/ability-to-update-or-create-an-event-without-sending-a/idi-p/2335070
  * https://learn.microsoft.com/en-us/answers/questions/1074414/create-an-event-but-dont-send-invitations
  * https://learn.microsoft.com/en-us/answers/questions/997970/msgraph-api-event-creation-suppress-email-feature

The only way I see for solving the duplicate mail issue is disabling Cal.Com notification mails, if an event was created using the `office365calendar` app.

**Comment by ldpg-dev** [comment]
Also the description is poorly formatted. Again, I'm not sure why this is labeled as "Improvement" and not as "bug", this is certainly a bug.

**Comment by jbcameraguy** [comment]
Any updates on this? I really don't want to, but I'm going to have to migrate my entire team back to Calendly because of this issue. I work in Customer Onboardings, and we send an AI recorder into the meetings and give the video links to the customers to review later, but because the Online Meeting isn't checked, the recorder isn't scheduled to enter the meeting and we have to manually send it in, which also doesn't auto trigger the video to send to the customer later. It's becoming a major problem.

**Comment by buzzard10** [comment]
Yeah, Microsoft Teams become a really popular service in COVID, and we're using this every day. Unfortunately, based on my last tests i can say that:
1. It doesn't generate link to meeting automatically
2. Calcom doesn't support webhook MEETING END for Microsoft Teams.

**Comment by herzkerl** [comment]
How about adding an option not to provide Microsoft's API with the attendee's email addresses? We also don't like the current behavior (of MS, that is!), but couldn't find any other workaround yet.

**Comment by tobiedesgreniers** [comment]
Hi guys, are there any update on this?
An opened PR? Priorities set? An investigation?
Is it going to be fix in the next few months or should we migrate to another scheduling platform?

**Comment by SecureCPU** [comment]
> Hi guys, are there any update on this? An opened PR? Priorities set? An investigation? Is it going to be fix in the next few months or should we migrate to another scheduling platform?

Unfortunately we've already had to migrate to another platform. Looking forward to when this is resolved to possibly migrate back. But this issue has been here for a very long while and there have been no updates. It completely messes us anyone using teams & outlook. Which these days is a lot of people & companies, so idk why it's taking so long to even give an update.

**Comment by johan-cruxdata** [comment]
+1 for priority on a resolution here. This is frustrating

**Comment by buzzard10** [comment]
Anyone have any workaround? I have to go to each meeting separately and switch this checkbox to make it an online meeting. Then it sends updates to all invited contacts. It's frustrating.

**Comment by wupdich** [comment]
This is such a big problem, the bad implementation and the duplicate emails. I can‘t recommend cal.com to my clients because of that. 
**Workaround** that I internally use right now to at least avoid duplicate emails is to add the event to a google calendar. Then I get automatically invited on my Outlook Calendar and the customer also just gets one email.

**Comment by emillod** [comment]
I believe I've resolved the issue on my end. After debugging, I discovered on the [Permissions page](https://learn.microsoft.com/en-us/graph/api/application-post-onlinemeetings?view=graph-rest-1.0&tabs=http#permissions) that the `OnlineMeetings.ReadWrite` permission must be enabled for the application. Previously, my Microsoft application only had one permission: `User.Read`. Here’s what I did to fix it:

1. Opened [Azure App Registration](https://portal.azure.com/#blade/Microsoft_AAD_IAM/ActiveDirectoryMenuBlade/RegisteredApps) and selected the Cal.com application.  
2. Clicked on **Manage** and navigated to **API permissions**.  
3. Clicked on **Add a permission**, then selected **Microsoft Graph**.  
4. Searched for and added `OnlineMeetings.ReadWrite`.  

After that, I re-added the Microsoft Teams integration to my Cal.com app, and it worked! This fix applies to both personal and team meetings. While I can’t guarantee how long it will remain functional, it’s now correctly adding the meeting link to the email body.
![image](https://github.com/user-attachments/assets/12dfe9f6-a657-4dd6-9c61-c5c382be1658)

**Comment by ldpg-dev** [comment]
@emillod Nice find! How does the email sent by cal and teams looks like? Does it also look the same as in the OP? Or was it also fixed?

**Comment by emillod** [comment]
It still poorly formatted and i customer receives two emails, one is nice from calcom and second one is from microsoft i believe and it's like a clean body. Maybe we can disable this in [API request](https://github.com/calcom/cal.com/blob/main/packages/app-store/office365video/lib/VideoApiAdapter.ts).
![image](https://github.com/user-attachments/assets/e90e780d-1d26-459b-8c83-6fd76de21dcf)

**Comment by robinmanuelthiel** [comment]
It's crazy, that this issue doesn't even have a **bug** label. Maybe we should rename it to the bug it is? @ldpg-dev

**Comment by PeerRich** [comment]
i heard from the team this is due to Microsoft being Microsoft and there isnt much we can do here

is this true?

**Comment by Rutik7066** [comment]
> i heard from the team this is due to Microsoft being Microsoft and there isnt much we can do here
> 
> is this true?

I am setting up the project. I haven't started to debug. If what you have heard is true, then yes, we might not be able to do anything

**Comment by anikdhabal** [comment]
Hey everyone, just came here from a linked PR. I have a few questions:

1. There’s a lot of discussion in this issue — could someone clarify what the actual issue is?
2. Is the issue currently reproducible? If not, I suggest we consider closing it.
3. Why does the issue title mention MS Teams? From what I see, it seems to be related to duplicate emails or something else.

**Comment by herzkerl** [comment]
When you save events in Microsoft 365 calendar, MS will send an e-mail to the participant(s). Since we don't want to communicate our personal e-mail addresses to customers (we only use them for authentication and some internal things), we'd like to turn this off somehow. I also tried this with NextCloud (via CalDAV integration), where this is _not_ the case.

I don't think it has anything to do with Teams, though—I just disabled my Teams integration, but still got the e-mail from Microsoft.

**Comment by vineeth-0509** [comment]
is this issue is still open,

**Comment by vijayraghav-io** [comment]
PR is created with a proper fix (with so much hardwork) with complete detailed testing, videos and approach laid out here with minute details https://github.com/calcom/cal.com/pull/21377.
This should be right fix - https://github.com/calcom/cal.com/pull/21377

**Comment by anikdhabal** [comment]
Hey @vijayraghav-io , thanks for your hard work. I noticed that your fix will break things for all active MS Teams users, so I decided not to proceed with that approach. That said, some parts of your code changes look good, and I'm considering giving you a chance to revise the fix and address the issue.

Reponeded it and looking forward to it 🙏

**Comment by vijayraghav-io** [comment]
Ok sure,  Thank you! Will update accordingly.
I had thought of this (existing MSteams users) but my approach or point of view is - user might not hesitate to reinstall (a one time step) for a better experience, because currently even reschedule or updateMeeting() also doesn't work as 'uid' or 'id' itself is not used in office365video/videoApiadapter.ts

**Comment by ldpg-dev** [comment]
@herzkerl What you described is actually **not the issue we're discussing here**. The issue is not the fact an event is sent from MS Teams with your MS Teams email exposed. That's to be expected, since the event is created using your MS Teams account, this is unavoidable, if you don't want your MS Teams account exposed, you can use Jitsi... In fact, the same behavior is true even with Calendly **which is good**. I believe this can be a separate Feature (since it's not a Bug) after the original problem with the integration is actually solved.

@vijayraghav-io @anikdhabal @PeerRich **Regardless of the many messages in this thread, the Bug we're all bumping up throughout the discussion is the one described in the original post:**


### Steps to Reproduce

1. Install MS Teams app from the app marketplace
2. Setup MS Outlook as your main Calendar
3. Create a Booking and set MS Teams as the meeting software
4. Book a meeting
5. Atendees will receive 2 emails: 1 from Cal.com + 1 from MS Calendar. It's fine and expected that they get two emails.

### Actual Results

**The problem is the one from MS Calendar looks like this:** _Notice the unprofessional description and formatting_

<img width="919" alt="image" src="https://github.com/calcom/cal.com/assets/81204440/f07a8f6b-b52c-4241-a501-a4d9e8cb1eeb">


**When I go to MS Teams to check, this is what I see:** _Notice I won't be able to join the meeting from the MS Teams App without copying and pasting the link in the browser since "Online meeting" is not checked + There is an invalid email address_

<img width="644" alt="image" src="https://github.com/calcom/cal.com/assets/81204440/dd5d9f82-cc75-4ab6-b3be-878c0058c564">


### Expected Results

This is what the MS Calendar email would normally look like when I create a meeting through Calendly or MS Teams directly:

<img width="557" alt="image" src="https://github.com/calcom/cal.com/assets/81204440/a3185eb4-a9f8-4089-9f0c-0bc4ea75c9a0">


Notice how the location does not need to be the Meeting Link but can be an arbitrary URL, in this case Calendly's event link, and the "Online Meeting" switch is turned on, letting Microsoft handle the link creation, instead of creating a link and then attaching it to the calendar event which is what Cal.com does (Not good)

<img width="632" alt="image" src="https://github.com/calcom/cal.com/assets/81204440/03350cdf-01e6-42a1-b104-b7c0cdc3965f">

**Comment by vijayraghav-io** [comment]
@ldpg-dev, yes the main reported in issue description is resolved by https://github.com/calcom/cal.com/pull/21377

**Comment by toluwagbemiga** [comment]
is this bounty still up?

**Comment by buzzard10** [comment]
Is this will also send a webhook after meeting ends?

**Comment by toluwagbemiga** [comment]
can i work on this
kindly respond
